inc/functions_upload.php

```diff
@@ -679,6 +679,83 @@ function upload_attachment($attachment, $update_attachment=false)
 	return $ret;
 }
 
+/**
+ * Process adding attachment(s) when the "Add Attachment" button is pressed.
+ *
+ * @param int $pid The ID of the post.
+ * @param array $forumpermission The permissions for the forum.
+ * @param string $attachwhere Search string "pid='$pid'" or "posthash='".$db->escape_string($mybb->get_input('posthash'))."'"
+ * @param string $action Where called from: "newthread", "newreply", or "editpost"
+ *
+ * @return array Array of errors if any, empty array otherwise
+ */
+function add_attachments($pid, $forumpermissions, $attachwhere, $action=false)
+{
+	global $db, $mybb, $editdraftpid, $lang;
+
+	$ret = array();
+
+	if($forumpermissions['canpostattachments'])
+	{
+		$attachments = array();
+		$fields = array ('name', 'type', 'tmp_name', 'error', 'size');
+
+		$total = isset($_FILES['attachments']['name']) ? count($_FILES['attachments']['name']) : 0;
+
+		for($i=0; $i<$total; ++$i)
+		{
+			foreach($fields as $field)
+			{
+				$attach1[$field] = $_FILES['attachments'][$field][$key];
+				$attachments[$i][$field] = $_FILES['attachments'][$field][$i];
+			}
+		}
+
+		foreach($attachments as $FILE)
+		{
+			if(!empty($FILE['name']) && !empty($FILE['type']))
+			{
+				if($FILE['size'] > 0)
+				{
+					$query = $db->simple_select("attachments", "aid", "filename='".$db->escape_string($FILE['name'])."' AND {$attachwhere}");
```

Comment:
Running a query per file doesn't seem like the best approach to me - would be nice to have one query with `WHERE filename IN (filenames...)` and then you could build an array of `[filename => aid]`.

Reply:
I fixed it, I doubt however if it is useful, as one query per attachment is little compared to 4-5 queries generated by `upload_attachment`

Reply:
Nice one, thanks. Anywhere where we can reduce the number of queries easily is worth doing IMO.

I'll give this a proper test tomorrow after work and hopefully merge. Thanks for all the hard work!